[CodeGen] Register-coalescer remat fix subreg liveness

VigneshwarJ opened the pull request:
This is a bugfix in rematerialization where the liveness of subreg mask was incorrectly updated causing crash in scheduler.

VigneshwarJ committed:
[CodeGen] Register-coalescer remat fix subreg liveness

This is a bugfix in rematerialization where the liveness of subreg mask
was incorrectly updated causing crash in scheduler.

VigneshwarJ committed:
fix tests

uweigand reviewed (commented):
SystemZ test change is OK.

arsenm commented on llvm/test/CodeGen/AMDGPU/reg-coalescer-subreg-liveness.mir:
```diff
@@ -0,0 +1,55 @@
+    undef %20.sub0:sgpr_128 = COPY %7
+    %0:sgpr_128 = COPY %20
+    %0.sub1:sgpr_128 = COPY killed %10
+    %27:sgpr_128 = COPY %0
```
Can you compact the register numbers (i.e., pass through -run-pass=none)

arsenm reviewed (commented):
lgtm with test cleanups

VigneshwarJ committed:
review comments

arsenm reviewed (changes requested):
On second look modifying the mask of the subrange doesn't look like a well formed transform

VigneshwarJ commented on llvm/test/CodeGen/AMDGPU/reg-coalescer-subreg-liveness.mir:
```diff
@@ -0,0 +1,58 @@
+  ; CHECK-NEXT: bb.1:
+  ; CHECK-NEXT:   successors: %bb.2(0x80000000)
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   S_NOP 0, implicit [[S_LOAD_DWORD_IMM]], implicit [[S_MOV_B32_]]
```
Yeah, I agree we don't want to introduce new defs in previously undefs. Yes, this is a deeper issue that this patch is not addressing. I will modify the mir to show the issue it is addressing.

VigneshwarJ replied:
I am looking at this de-optimizing as a separate bug from our discussion. The point to note is that the register-coalescer was generating the exact code before this patch, too; only the lanemasks were corrupt.

arsenm replied:
The S_NOP disappeared in the latest version of the patch? Did refineSubRanges fix this?

arsenm replied:
What's the difference? Can you include both versions?

VigneshwarJ replied:
The first commit had S_NOP (with implicit to just verify liveness), which was a reduced case.  I thought that was not a clear test case.  In the actual bug, there was a tensor_load instruction that took a sgpr128 with all four sublanes.

VigneshwarJ commented on llvm/lib/CodeGen/RegisterCoalescer.cpp:
This comment mentions the same, but it is not being done right when there's an overlap in  (SR.LaneMask & DstMask)

VigneshwarJ commented:
the mir I wrote was ambiguous and as mentioned above has other problems. Updated the mir test.
The issue this ticket tries to address is
```llvm
48B	  %2:sreg_32 = S_MOV_B32 1
64B	  undef %3.sub0:sgpr_128 = COPY %2:sreg_32
..
128B	  %6:sgpr_128 = COPY killed %3:sgpr_128 (lanemask 0xF3  [128r, 464B:0))
144B	  %6.sub1:sgpr_128 = COPY killed %1:sreg_32_xm0_xexec (lanemask 0x0C [128r,128d:1))
..
464B	bb.2:
```

Here the coalescer remats to undef %3.sub0 = S_MOV_B32 1
And when it tries to remat 128B
		updated: 128B	undef %6.sub0:sgpr_128 = S_MOV_B32 1
		updated: 144B	%6.sub1:sgpr_128 = COPY %1:sreg_32_xm0_xexec
But the lanemask is still previous (lanemask 0xF3  [128r, 464B:0)) . This causes 

Because this is a new def with undef for subreg , the lanemask should be  (lanemask 0x03  [128r, 464B:0)), this is what this ticket fixes.

VigneshwarJ committed:
updated tests

arsenm commented on llvm/lib/CodeGen/RegisterCoalescer.cpp:
```diff
@@ -1625,6 +1625,7 @@ bool RegisterCoalescer::reMaterializeDef(const CoalescerPair &CP,
           // dead def so that the interferences are properly modeled.
           if (!SR.liveAt(DefIndex))
             SR.createDeadDef(DefIndex, Alloc);
+          SR.LaneMask = DstMask & SR.LaneMask;
```
I do not think you should be trying to mutate the LaneMask inside the loop, and only operations on the main range should modify the lane masks of the contained subranges. I think you need to do something outside of the loop, e.g. something like refineSubRanges

VigneshwarJ committed:
changed to refineSubrange

VigneshwarJ committed:
added one more test
